### PR TITLE
Cubeprinter adopt

### PR DESCRIPTION
**Pull request by pp-mo**
### Description
Followon from #3998 
Final small formatting changes
Unsurprisingly, I had to squash + rebase from #3998 to get this testing.

Closes : #4211
---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by pp-mo** [comment]
Surprisingly, this caused **no** obvious problems.
So now I'm now going to crowbar-replace the old with the new styles, and see what happens ...

**Comment by pp-mo** [comment]
22 failures is really not that bad.
I think, at that rate, that I will simply adjust the reference files.
 -- whereas I was considering a 'clever' testing strategy more akin to the 'tolerant comparison' techniques used above.
From discussion with @bjlittle I don't think we need that -- simpler to just change the references.
Likewise, decided to remove the additional colons in scalar-coord and attribute outputs.

**Comment by pp-mo** [comment]
I don't understand the latest docstest fail.
When I run the doctests locally, it insists that the result of printing a `cube.metadata._asdict()` shows as an OrderedDict.
And I can confirm that it looks like that, e.g.
```
>>> print(Cube([0]).metadata._asdict())
OrderedDict([('standard_name', None), ('long_name', None), ('var_name', None), ('units', Unit('unknown')), ('attributes', {}), ('cell_methods', ())])
>>> 
```

For me it's doing that in my test env too, and with older versions of Iris.

But in the CI testing it prints out as an ordinary dict.
And that is in spite of my running the tests via nox : `nox -s doctest-3.7` -- which really ought to take care of any environment differences ??
Possibly it is a lockfiles thing ?

**UPDATE:**
It is a Python 37/38 difference :
```
$ conda activate py37
$ python -c "import sys; print(sys.version); from collections import namedtuple; nt = namedtuple('nt', ['a','b']); print(nt(a=1, b=2)._asdict())"
3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
[GCC 9.3.0]
OrderedDict([('a', 1), ('b', 2)])
$                    
$ conda activate py38
$ python -c "import sys; print(sys.version); from collections import namedtuple; nt = namedtuple('nt', ['a','b']); print(nt(a=1, b=2)._asdict())"
3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05) 
[GCC 9.3.0]
{'a': 1, 'b': 2}
$
```

**Comment by pp-mo** [comment]
Yay, done at last.
@bjlittle very grateful for your eyes on this ..

File: lib/iris/tests/unit/experimental/representation/test_CubeRepresentation.py [hunk 1]
```diff
@@ -354,14 +354,6 @@ def test_not_included(self):
         for heading in not_included:
             self.assertNotIn(heading, self.result)
 
-    def test_handle_newline(self):
```
Comment (pp-mo) [inline lib/iris/tests/unit/experimental/representation/test_CubeRepresentation.py:357]: NOTE: we simply don't output embedded newlines in the "new world" : we quote the string instead.
So this is no longer needed

**Comment by trexfeathers** [comment]
>Followon from #3998

Does this supercede #3998?

**Comment by pp-mo** [comment]
> Does this supercede #3998?

Absolutely, yes.
( Sorry, sure I posted this earlier but it vanished? perhaps failed to hit the button )

**Comment by pp-mo** [comment]
Thanks for looking @bjlittle :eyes: 
Just now spotted  conflicts + fixing that  ...

**Comment by pp-mo** [comment]
Hit that spurious, intermittent failure of the animation test again.
Now fixed.

Unfortunately I re-spun it and so have lost the reference.
But there is a problem lurking there somewhere ...

**Comment by pp-mo** [comment]
Merged in the re-ordering from #4251 that ensures attributes come last, after cell-methods.
Fixed resulting test changes.
All passing now.
@bjlittle ready when you are !

**Review by bjlittle — CHANGES_REQUESTED** [review]
@pp-mo Great stuff :+1: 

Thanks for diligently wading through this unpleasant mire of `Cube.summary` refactorisation, the result is a vast improvement IMHO.

Just a couple of minor comments to service, then we can bank it :beers:

File: lib/iris/_representation/cube_printout.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        else:
+            cube_summary = CubeSummary(cube_or_summary)
+        self.table = self._ingest_summary(cube_summary)
+
+    def _ingest_summary(
+        self,
+        cube_summary,
+        n_indent_section=4,
+        n_indent_item=4,
+        n_indent_extra=4,
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py:174]: Agreed. With this particular modularity I see no prospect of re-use by a different caller. Will address...

File: lib/iris/_representation/cube_printout.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    ):
+        """Make a table of strings representing the cube-summary."""
+        sect_indent = " " * n_indent_section
+        item_indent = sect_indent + " " * n_indent_item
+        item_to_extra_indent = " " * n_indent_extra
+        extra_indent = item_indent + item_to_extra_indent
+        summ = cube_summary
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py:181]: I'm going to stop assigning it.

File: docs/src/further_topics/metadata.rst [hunk 4]
```diff
@@ -108,22 +108,22 @@ For example, given the following :class:`~iris.cube.Cube`,
...
+            time                             x              -              -
+            latitude                         -              x              -
+            longitude                        -              -              x
+        Auxiliary coordinates:
+            forecast_period                  x              -              -
+        Scalar coordinates:
+            forecast_reference_time     1859-09-01 06:00:00
+            height                      1.5 m
+        Cell methods:
+            mean: time (6 hour)
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply docs/src/further_topics/metadata.rst]: 28f3fb05da8ec34b1fd8a9a6be7c13d82badf24c

File: lib/iris/tests/results/cdm/str_repr/cell_methods.__str__.txt [hunk 5]
```diff
@@ -1,17 +1,17 @@
...
+        longitude                                                                                                             -              x
+    Scalar coordinates:
+        forecast_period                                                                                              6477.0 hours
+        forecast_reference_time                                                                                      1998-03-06 03:00:00
+        pressure                                                                                                     1000.0 hPa
+        time                                                                                                         1998-12-01 00:00:00
+    Cell methods:
+        mean: longitude (6 minutes, This is a test comment), latitude (12 minutes)
+        average: longitude (6 minutes, This is another test comment), latitude (15 minutes, This is another comment)
+        average: longitude, latitude
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/tests/results/cdm/str_repr/cell_methods.__str__.txt]: 28f3fb05da8ec34b1fd8a9a6be7c13d82badf24c

File: lib/iris/_representation/cube_printout.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+class CubePrinter:
+    """
+    An object created from a
+    :class:`iris._representation.cube_or_summary.CubeSummary`, which provides
+    text printout of a :class:`iris.cube.Cube`.
+
+    TODO: the cube :meth:`iris.cube.Cube.__str__` and
+    :meth:`iris.cube.Cube.__repr__`  methods, and
+    :meth:`iris.cube.Cube.summary` with 'oneline=True', should use this to
+    produce cube summary strings.
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py]: 6fee9977e99d578b24f305e32327cd5e58d0c798

File: lib/iris/_representation/cube_printout.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+        return result_lines
+
+    def __str__(self):
+        return "\n".join(self.formatted_as_strings())
+
+
+class CubePrinter:
+    """
+    An object created from a
+    :class:`iris._representation.cube_or_summary.CubeSummary`, which provides
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py]: 3f927e7d62167e91459abd186d6445ecb251ec5b

File: lib/iris/_representation/cube_printout.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+    text printout of a :class:`iris.cube.Cube`.
+
+    TODO: the cube :meth:`iris.cube.Cube.__str__` and
+    :meth:`iris.cube.Cube.__repr__`  methods, and
+    :meth:`iris.cube.Cube.summary` with 'oneline=True', should use this to
+    produce cube summary strings.
+
+    This class has no internal knowledge of :class:`iris.cube.Cube`, but only
+    of :class:`iris._representation.cube_or_summary.CubeSummary`.
```
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py]: 3f927e7d62167e91459abd186d6445ecb251ec5b

File: lib/iris/_representation/cube_printout.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py]: fe3dadbdea954e4f085e009ce20a9b2446da0c29

File: lib/iris/_representation/cube_printout.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trexfeathers) [reply lib/iris/_representation/cube_printout.py]: b6c19ff605d0968bb90ce82aaba69676e4e70bb3

File: lib/iris/_representation/cube_summary.py [hunk 9]
```diff
@@ -218,7 +218,18 @@ def __init__(self, title, attributes):
 class CellMethodSection(Section):
     def __init__(self, title, cell_methods):
         self.title = title
-        self.contents = [str(cm) for cm in cell_methods]
+        # self.contents = [str(cm) for cm in cell_methods]
```
Comment (bjlittle) [inline lib/iris/_representation/cube_summary.py:221]: @trexfeathers Oops... you forgot to nuke this `wip` code comment :+1:

File: docs/src/further_topics/metadata.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (bjlittle) [reply docs/src/further_topics/metadata.rst]: This works out just lovely now, nice one :100:

File: lib/iris/_representation/cube_summary.py (hunk 9 above)
Comment (trexfeathers) [reply lib/iris/_representation/cube_summary.py]: d4843fc805e15c38f9510621c76136764aa8ef05

**Review by bjlittle — CHANGES_REQUESTED** [review]
@trexfeathers Thanks for nudging this PR across the line, it's really appreciated.

Just one last comment to service, plus could you add a `whatsnew` entry too?

Make sure to attribute it to both @pp-mo and yourself for your awesome combined efforts :beers: 

Also, since this PR is quite noisy, I'm going to wait until #4276 lands before merging this PR... as #4276 was a tad involved and I think it would be easier for this PR to be rebased than for #4276 to fix merge conflicts.

**Comment by trexfeathers** [comment]
The what's new is conflicted. I'll hold off rebasing since @bjlittle anticipates more conflicts post #4276.

**Pull request merged**